frame/bridge/ethereum/issuing/src/lib.rs

```diff
@@ -21,275 +21,351 @@
+	ensure,
+	pallet_prelude::*,
+	parameter_types,
 	traits::{Currency, ExistenceRequirement::*, Get},
-	weights::Weight,
 	PalletId,
 };
 use frame_system::{ensure_root, ensure_signed, pallet_prelude::*};
 use sp_runtime::{
-	traits::{AccountIdConversion, Saturating},
+	traits::{AccountIdConversion, Keccak256, Saturating},
 	AccountId32, DispatchError, DispatchResult, SaturatedConversion,
 };
-use sp_std::vec::Vec;
+use sp_std::{str, vec::Vec};
 // --- darwinia ---
-use darwinia_ethereum_issuing_contract::{
-	Abi, Event as EthEvent, Log as EthLog, TokenBurnInfo, TokenRegisterInfo,
-};
-use darwinia_evm::{GasWeightMapping, IssuingHandler};
+use darwinia_evm::AddressMapping;
+use darwinia_evm::GasWeightMapping;
 use darwinia_relay_primitives::relay_authorities::*;
-use darwinia_support::{balance::*, traits::EthereumReceipt};
-use dp_evm::CallOrCreateInfo;
+use darwinia_support::{balance::*, traits::EthereumReceipt, PalletDigest};
+use dp_contract::{
+	ethereum_backing::{EthereumBacking, EthereumLockEvent, EthereumRegisterEvent},
+	mapping_token_factory::{
+		MappingTokenFactory as mtf, TokenBurnInfo, TokenRegisterInfo, BURN_ACTION, REGISTER_ACTION,
+	},
+};
 use ethereum_primitives::{
 	receipt::{EthereumTransactionIndex, LogEntry},
-	EthereumAddress,
+	EthereumAddress, U256,
 };
-use types::*;
-
-pub trait Config: dvm_ethereum::Config {
-	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
-
-	type PalletId: Get<PalletId>;
-
-	type RingCurrency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
-
-	type EthereumRelay: EthereumReceipt<Self::AccountId, RingBalance<Self>>;
-	type EcdsaAuthorities: RelayAuthorityProtocol<Self::BlockNumber, Signer = EthereumAddress>;
-
-	type WeightInfo: WeightInfo;
-}
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	mod types {
+		use crate::*;
+
+		pub type AccountId<T> = <T as frame_system::Config>::AccountId;
+		pub type EthereumReceiptProofThing<T> = <<T as Config>::EthereumRelay as EthereumReceipt<
+			AccountId<T>,
+			RingBalance<T>,
+		>>::EthereumReceiptProofThing;
+		pub type RingBalance<T> = <<T as Config>::RingCurrency as Currency<AccountId<T>>>::Balance;
+	}
+	pub use types::*;
+
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: dvm_ethereum::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+		type RingCurrency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+		type EthereumRelay: EthereumReceipt<Self::AccountId, RingBalance<Self>>;
+		type EcdsaAuthorities: RelayAuthorityProtocol<Self::BlockNumber, Signer = EthereumAddress>;
+		type WeightInfo: WeightInfo;
+	}
 
-decl_error! {
+	#[pallet::error]
 	/// Issuing pallet errors.
-	pub enum Error for Module<T: Config> {
+	pub enum Error<T> {
 		/// Invalid Issuing System Account
 		InvalidIssuingAccount,
 		/// assert ar
 		AssetAR,
```

Comment:
`AR` means `Assert Already Registered`? Maybe we should give a more understandable name.